### PR TITLE
Fork environment

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: src/pipe.toit [hunk 1]
```diff
@@ -165,9 +165,20 @@ To avoid zombies you must either give the child process to either
...
   child process as open file descriptors 3 and/or 4.
+Optionally, $environment variables can be passed as a map.
+  Keys in the map should be strings, and values should be strings or null,
+  where null indicates that the variable should be unset in the child
+  process.
+Note that if you override the PATH environment variable, but set the $use_path
+  flag, the new value of PATH will be used to find the executable.
 */
-fork use_path stdin stdout stderr --file_descriptor_3/OpenPipe?=null --file_descriptor_4/OpenPipe?=null command arguments -> List:
+fork use_path stdin stdout stderr --environment/Map?=null --file_descriptor_3/OpenPipe?=null --file_descriptor_4/OpenPipe?=null command arguments -> List:
```
Comment (floitsch) [inline src/pipe.toit:175]: Split into multiple lines?

File: src/pipe.toit [hunk 2]
```diff
@@ -165,9 +165,20 @@ To avoid zombies you must either give the child process to either
...
   result := List 4
+  flat_environment := environment ? (Array_ environment.size * 2) : null
+  index := 0
+  if environment: environment.do: | key value |
+    flat_environment[index++] = key.stringify
+    flat_environment[index++] = value ? value.stringify : null
```
Comment (floitsch) [inline src/pipe.toit:181]: What if `value` is 'false' ?
or do you only care for `null` here?
if it's only `null` then you might as well write `value and value.stringify`.

File: tests/echo.toit [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        var = part[1..close]
+        rest = part[close + 1 ..]
+      else:
+        for i := 0; i < part.size; i++:
+          c := part[i]
+          if not is_identifier c:
+            var = part[0..i]
+            rest = part[i..]
+            break
+      catch: result.add os.env[var]
```
Comment (floitsch) [inline tests/echo.toit:36]: Maybe use `os.env.get`?

File: tests/echo.toit [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+import host.os
+
+// Expands environment variables in the arguments, using $ syntax.
+main args:
+  print
+      (args.map: expand it).join " "
+
+expand src -> string:
+  dont_expand := true
```
Comment (floitsch) [inline tests/echo.toit:13]: change to `expand := false` ?

File: tests/echo.toit [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// be found in the tests/LICENSE file.
+
+import host.os
+
+// Expands environment variables in the arguments, using $ syntax.
+main args:
+  print
+      (args.map: expand it).join " "
+
+expand src -> string:
```
Comment (floitsch) [inline tests/echo.toit:12]: Given an example of what this function does.

File: tests/echo.toit [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  dont_expand := true
+  result := []
+  src.split "\$": | part |
+    if dont_expand:
+      result.add part
+      dont_expand = false
+    else if part == "":
```
Comment (floitsch) [inline tests/echo.toit:19]: Add comment explaining that `$$` evaluates to `$`.

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/pipe.toit (hunk 2 above)
Comment (erikcorry) [reply src/pipe.toit:181]: changed to: (value == null) ? null : value.stringify

which will pass false as a string, not as a bool.

**Pull request merged**